Tests fail with go1.19

Several tests fail with go1.19, but pass with go1.18:

```console
$ go version
go version go1.19 linux/amd64
$ go test ./...
--- FAIL: TestGetKey (2.99s)
    gokey_test.go:161: keys with same invocation options do not match
FAIL
FAIL	github.com/cloudflare/gokey	3.153s
?   	github.com/cloudflare/gokey/cmd/gokey	[no test files]
--- FAIL: TestKnownKey (0.20s)
    keygen_test.go:129: generated RSA 2048 does not match the expected result
--- FAIL: TestStdKey (0.25s)
    keygen_test.go:149: mitigating crypto/internal/randutil.MaybeReadByte...
    keygen_test.go:149: mitigating crypto/internal/randutil.MaybeReadByte...
    keygen_test.go:194: RSA key generation algorithm from stdlib deviates from the one in gokey
FAIL
FAIL	github.com/cloudflare/gokey/rsa	0.454s
FAIL
```

See the Debian FTBFS (fail-to-build-from-source) bug at https://bugs.debian.org/1017300 where @lnussbaum routinely rebuilds packages with the latest tools and libraries, and in this case, go1.19.  It probably has to do with changes made to go1.19 built-in crypto libraries, see https://go.dev/doc/go1.19#minor_library_changes

Many thanks!

--- a/rsa/keygen.go
+++ b/rsa/keygen.go
@@ -1,13 +1,107 @@
 package rsa
 
 import (
-	"crypto/rand"
 	"crypto/rsa"
 	"errors"
 	"io"
 	"math"
 	"math/big"
 )
+
+// below is taken from https://github.com/golang/go/blob/8a2553e380196dda556608e2fe79881004770eb9/src/crypto/rand/util.go#L31
+// because later versions of rand.Prime() have a different algorithm and are not deterministic
+
+// smallPrimes is a list of small, prime numbers that allows us to rapidly
+// exclude some fraction of composite candidates when searching for a random
+// prime. This list is truncated at the point where smallPrimesProduct exceeds
+// a uint64. It does not include two because we ensure that the candidates are
+// odd by construction.
+var smallPrimes = []uint8{
+	3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53,
+}
+
+// smallPrimesProduct is the product of the values in smallPrimes and allows us
+// to reduce a candidate prime by this number and then determine whether it's
+// coprime to all the elements of smallPrimes without further big.Int
+// operations.
+var smallPrimesProduct = new(big.Int).SetUint64(16294579238595022365)
+
+// prime returns a number, p, of the given size, such that p is prime
+// with high probability.
+// prime will return error for any error returned by rand.Read or if bits < 2.
+func prime(rand io.Reader, bits int) (p *big.Int, err error) {
+	if bits < 2 {
+		err = errors.New("crypto/rand: prime size must be at least 2-bit")
+		return
+	}
+
+	b := uint(bits % 8)
+	if b == 0 {
+		b = 8
+	}
+
+	bytes := make([]byte, (bits+7)/8)
+	p = new(big.Int)
+
+	bigMod := new(big.Int)
+
+	for {
+		_, err = io.ReadFull(rand, bytes)
+		if err != nil {
+			return nil, err
+		}
+
+		// Clear bits in the first byte to make sure the candidate has a size <= bits.
+		bytes[0] &= uint8(int(1<<b) - 1)
+		// Don't let the value be too small, i.e, set the most significant two bits.
+		// Setting the top two bits, rather than just the top bit,
+		// means that when two of these values are multiplied together,
+		// the result isn't ever one bit short.
+		if b >= 2 {
+			bytes[0] |= 3 << (b - 2)
+		} else {
+			// Here b==1, because b cannot be zero.
+			bytes[0] |= 1
+			if len(bytes) > 1 {
+				bytes[1] |= 0x80
+			}
+		}
+		// Make the value odd since an even number this large certainly isn't prime.
+		bytes[len(bytes)-1] |= 1
+
+		p.SetBytes(bytes)
+
+		// Calculate the value mod the product of smallPrimes. If it's
+		// a multiple of any of these primes we add two until it isn't.
+		// The probability of overflowing is minimal and can be ignored
+		// because we still perform Miller-Rabin tests on the result.
+		bigMod.Mod(p, smallPrimesProduct)
+		mod := bigMod.Uint64()
+
+	NextDelta:
+		for delta := uint64(0); delta < 1<<20; delta += 2 {
+			m := mod + delta
+			for _, prime := range smallPrimes {
+				if m%uint64(prime) == 0 && (bits > 6 || m != uint64(prime)) {
+					continue NextDelta
+				}
+			}
+
+			if delta > 0 {
+				bigMod.SetUint64(delta)
+				p.Add(p, bigMod)
+			}
+			break
+		}
+
+		// There is a tiny possibility that, by adding delta, we caused
+		// the number to be one bit too long. Thus we check BitLen
+		// here.
+		if p.ProbablyPrime(20) && p.BitLen() == bits {
+			return
+		}
+	}
+}
 
 // below is taken from https://github.com/golang/go/blob/161874da2ab6d5372043a1f3938a81a19d1165ad/src/crypto/rsa/rsa.go#L220
 // because later versions in standard crypto/rsa package are not deterministic
@@ -61,7 +155,7 @@ NextSetOfPrimes:
 		}
 		for i := 0; i < nprimes; i++ {
 			var err error
-			primes[i], err = rand.Prime(random, todo/(nprimes-i))
+			primes[i], err = prime(random, todo/(nprimes-i))
 			if err != nil {
 				return nil, err
 			}

--- a/rsa/keygen_test.go
+++ b/rsa/keygen_test.go
@@ -2,10 +2,7 @@ package rsa_test
 
 import (
 	"bytes"
-	"crypto/rsa"
 	"encoding/pem"
-	"io"
-	"runtime/debug"
 	"strings"
 	"testing"
 
@@ -132,70 +129,5 @@ func TestKnownKey(t *testing.T) {
 	rsa4096 := getKey(gokey.RSA4096, t)
 	if !pemEqual(rsa4096, knownRsa4096, t) {
 		t.Fatal("generated RSA 4096 does not match the expected result")
-	}
-}
-
-// reader which tries to negate the effect of crypto/internal/randutil/randutil.MaybeReadByte
-// https://github.com/golang/go/blob/6269dcdc24d74379d8a609ce886149811020b2cc/src/crypto/internal/randutil/randutil.go#L25
-// useful to regain deterministic output of some crypto routines with respect to a fixed pseudo-random stream
-// such as https://github.com/golang/go/blob/6269dcdc24d74379d8a609ce886149811020b2cc/src/crypto/rsa/rsa.go#L222
-type maybenotReader struct {
-	io.Reader
-	*testing.T
-}
-
-func (r maybenotReader) Read(p []byte) (n int, err error) {
-	if strings.Contains(string(debug.Stack()), "crypto/internal/randutil.MaybeReadByte") {
-		r.T.Log("mitigating crypto/internal/randutil.MaybeReadByte...")
-		// feed MaybeReadByte with dummy zeroes
-		for i := range p {
-			p[i] = 0
-		}
-		return len(p), nil
-	}
-
-	return r.Reader.Read(p)
-}
-
-// generate an RSA key using stdlib crypto/rsa package
-// but with crypto/internal/randutil.MaybeReadByte removed
-func getStdRsaKey(bits int, t *testing.T) string {
-	realm := "example.com"
-
-	// mimic gokey.GetKey behaviour
-	switch bits {
-	case 2048:
-		realm += "-key(RSA2048)"
-	case 4096:
-		realm += "-key(RSA4096)"
-	}
-
-	rng := &maybenotReader{gokey.NewDRNG("pass", realm), t}
-	priv, err := rsa.GenerateKey(rng, bits)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	var b strings.Builder
-	err = gokey.EncodeToPem(priv, &b)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	return b.String()
-}
-
-// this test is here to see if the RSA key generation algorithm in the stdlib is the same as in our package
-// for now we want to track any stdlib changes to the algorithm
-// however, this test might be removed in the future, if the changes in the stdlib will not be backwards compatible
-func TestStdKey(t *testing.T) {
-	rsa2048 := getStdRsaKey(2048, t)
-	if !pemEqual(rsa2048, knownRsa2048, t) {
-		t.Fatal("RSA key generation algorithm from stdlib deviates from the one in gokey")
-	}
-
-	rsa4096 := getStdRsaKey(4096, t)
-	if !pemEqual(rsa4096, knownRsa4096, t) {
-		t.Fatal("RSA key generation algorithm from stdlib deviates from the one in gokey")
 	}
 }

